amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShutdownTask.java

```diff
@@ -203,6 +209,31 @@ private void updateLeasesForChildShards()
         log.info("Shard {}: Updated current lease {} with child shard information: {}", shardInfo.shardId(), currentLease.leaseKey(), childShardIds);
     }
 
+    private void garbageCollectLeaseForParentShard() throws DependencyException, ProvisionedThroughputException, InvalidStateException {
+       final Lease parentShardLease = leaseCoordinator.getCurrentlyHeldLease(leaseKeyProvider.apply(shardInfo));
+
+        if (parentShardLease.childShardIds().isEmpty()) {
+            log.warn("No child shard state found in lease {} - not deleting lease.", parentShardLease);
+        } else {
+            final Set<String> expectedLeaseKeys = parentShardLease.childShardIds().stream()
+                    .map(c -> ShardInfo.getLeaseKey(shardInfo, c)).collect(Collectors.toSet());
+            final Set<String> leaseKeys = new HashSet<>();
+
+            for (String childShard : parentShardLease.childShardIds()) {
+                final String leaseKey = ShardInfo.getLeaseKey(shardInfo, childShard);
+                leaseKeys.add(leaseCoordinator.leaseRefresher().getLease(leaseKey).leaseKey());
+            }
+
+            if (Objects.equals(expectedLeaseKeys, leaseKeys)) {
+                log.debug("Finished processing lease {} - deleting lease.", parentShardLease);
+                leaseCoordinator.leaseRefresher().deleteLease(parentShardLease);
```

Comment:
We should only delete the parent lease/s when all the child lease/s have started processing. And if we need to introduce the deferred lease cleaning, we need to check how long the parent shard reached shard end. Have we got a conclusion of whether to introduce deferred lease cleaning?

Reply:
1. Deferred lease cleanup as discussed offline will not be added, since the edge case of a single worker recreating closed shard leases is very slim; requiring the customer to intentionally create holes in their lease table, use a high number of elected leaders, and have one worker which is extremely slow to use the stale shard info.

2. Yeah, will add this check for the child shards.